File: DataStructures/src/main/java/com/zhangdd/recusion/Queue8.java
package com.zhangdd.recusion;

import org.junit.Test;

/**
 * @author zhangdd on 2021/10/14
 */
public class Queue8 {

    final int MAX = 8;
    int[] arr = new int[MAX];
    int solutionCount = 0;


    @Test
    public void findSolution() {
        doPoint(0);
        System.out.printf("一个有%d中解法", solutionCount);
    }


    /**
     * 判断第n个皇后位置是否合法
     * <p>
     * 即判断和前n-1个是否在同一行、同一列、同一斜线
     */
    private boolean illegal(int n) {
        for (int i = 0; i < n; i++) {
            //(i,arr[i])  (n,arr[n])
            if (arr[i] == arr[n] || Math.abs(i - n) == Math.abs(arr[i] - arr[n])) {
                return false;
            }
        }
        return true;
    }

    /**
     * 放第n个皇后
     */
    private void doPoint(int n) {
        if (n == MAX) {
            showSolution();
            return;
        }
        for (int i = 0; i < MAX; i++) {
            arr[n] = i;
            if (illegal(n)) {
                doPoint(n + 1);
            }
        }
    }

    /**
     * 展示当前解法
     */
    private void showSolution() {
        solutionCount++;
        for (int i : arr) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

}

File: python/DataStructures/recusion/Queue8.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/2/3
八皇后问题
"""

MAX = 8


class Queue8:

    def __init__(self):
        self.position = [-1] * MAX
        self.count = 0

    '''
    先定一个长度大小为8的一维数组position，初始化值为-1，表示都还没放
    
    先放第一个皇后，第一个皇后肯定在第一行，也就是处理position[0]=? 
    尝试从第一行的第一列开始 即假设设置position[0]=0 ，看是否冲突，和在此之前放过的比较，由于是第一个放的，所以没有冲突
    
    开始放第二个皇后，第二个皇后肯定在第二行，也就是要处理position[1]=? 的问题
    尝试从第二行的第一列开始 即假设设置position[0]=0 ，看是否冲突，和在此之前放过的比较 ，发现有冲突，
    尝试从第二行的第二列开始 即假设设置position[0]=1 ，看是否冲突，和在此之前放过的比较 ，
    ......
    
    重要的当前这个位置如果是可以的话，进行递归操作 也就是保存了当前的栈情况
    '''

    def is_conflict(self, n):
        for i in range(n):
            if self.position[i] == self.position[n] or abs(n - i) == abs(self.position[n] - self.position[i]):
                return True
        return False

    def play(self):
        self.do_play(0)
        print('一共有{0}种解法'.format(self.count))

    def do_play(self, n):
        if n == MAX:
            self.count += 1
            self.show_solution()
            return
        for i in range(MAX):
            self.position[n] = i
            if not self.is_conflict(n):
                self.do_play(n + 1)

    def show_solution(self):
        print(self.position)


queue8 = Queue8()
queue8.play()
